### DISCUSSION TITLE
Discussion Around Scala Native 0.5 Upgrade

### DISCUSSION THREAD

**Discussion by djspiewak** [typelevel/governance#135] [General]
All topics related to Scala Native 0.5 upgrade. Relocating from Discord ([reference anchor](https://discord.com/channels/632277896739946517/920140670394589224/1229915431439958046)). Delegate issues can be collected here: https://github.com/orgs/typelevel/projects/2/views/1

**Comment by djspiewak** [comment]
One of the things we need to decide is whether we are going to just publish a bunch of snapshots first and get everything working before the very first release, or if we're comfortable releasing some very slow-moving upstream projects (such as ScalaCheck). The danger here is it's a one-way door. If we release a full ScalaCheck against 0.5, we can't go back again and release a new ScalaCheck against 0.4.x. This could be a problem if the downstream ecosystem (say, Cats Effect) takes a really long time to upgrade to 0.5 and ends up needing some new fix in ScalaCheck on an older stable lineage.

So the *safe* thing to do is just snapshot stuff, but ScalaCheck is very much the root of the Scala universe, so we would unblock more people by releasing at least it (and possibly also Discipline).

**Reply by armanbilge** [reply]
> if we're comfortable releasing some very slow-moving upstream projects (such as ScalaCheck). The danger here is it's a one-way door.

We should exercise caution, but I am comfortable with doing this for slow-moving, stable projects, provided that we do a couple things:
1. We do one last release on SN 0.4, so the existing ecosystem can get the latest features/fixes.
2. Then we bump a minor version and publish a release that _just_ upgrades to SN 0.5. This should unblock the emerging ecosystem and we can still release patches on the old series just-in-case.

**Reply by TonioGela** [reply]
So, which are the slow-moving libs? ScalaCheck and Discipline?
I'll leave this here just in case: https://x.com/mpilquist/status/1424881182437789702

**Reply by mpilquist** [reply]
I'd argue for all the way through cats and then stopping before cats-effect.

**Reply by mpilquist** [reply]
On second thought, we should consider publishing everything for 0.5 to unblock downstream libraries. Then we can come back and release snapshots / milestones of CE using the new multithreading support. @keynmol convinced me of this [on Discord](https://discord.com/channels/632277896739946517/920140670394589224/1230066686439456859).

**Reply by armanbilge** [reply]
I responded [on Discord](https://discord.com/channels/632277896739946517/920140670394589224/1230589678076428329). Basically my concern is that we might not want to make a CE release on SN 0.5 that supports single-threading at all, because this runtime is actually broken due to its inability to properly shift blocking tasks. Until we have a good case that continuing to support single-threaded compilation is valuable, I would prefer to go straight to multi-threading support with no support for single-threading.

Furthermore, properly supporting multi-threading means making changes to CE that can only happen in a breaking bump. ~~For example we can now restore the `Dispatcher#unsafeRunSync` API on Scala Native, but this a binary-breaking change. So we need to make it _before_ we publish anything for 0.5.~~ Sorry, that example is actually not correct. We can restore that API with a default implementation without breaking compatibility. However, there may be other examples like that.

**Reply by djspiewak** [reply]
While I agree in principle with Arman, the concern I have is that unless we get very lucky, we're not really close to having a multi-threaded SN runtime up and running, and if we get slightly-less-than-lucky it could be quite a long and drawn-out affair. This is particularly true if we couple it with the polling runtime, which I genuinely think we need to do. If we take a hard line and say that we're not going to release CE for 0.5 until we solve all those problems, the downstream ecosystem is in for a long wait.

Maybe that's okay! Maybe that's what everyone is signing up for, but it feels… implausible. I would *assume* that it's better to publish for 0.5, perhaps even in the 3.5.x series, and then relieve the time pressure on getting real multithreaded support out the door.

**Reply by mpilquist** [reply]
+1 for unblocking ecosystem first then following up with MT.

**Reply by armanbilge** [reply]
> This is particularly true if we couple it with the polling runtime

Right, well, I encouraged that we get the Native version of the polling runtime out in CE 3.5 so that we would avoid this situation of having to do polling+multithreading at the same time, but you pushed back on the principles despite the significant practical benefits it would offer 🙂

Idk. Seems to be just my opinion at this point 🙂 I am also disappointed that MUnit released for SN 0.5 without fixing any of the single-threaded assumptions baked in. It all feels very haphazard.

**Reply by mzuehlke** [reply]
Are you referring to https://github.com/scalameta/munit/issues/500 ?

**Reply by JD557** [reply]
I think it's a reference to some of the implementations in https://github.com/scalameta/munit/blob/main/munit/native/src/main/scala/munit/internal/PlatformCompat.scala. At least `waitAtMost` and `setTimeout` are not that great, even when running in a single thread.

Having said that, while some of the implementations there are a bit broken, I think they _mostly_ work, so having an early release to unblock the rest of the ecosystem might not have been that bad

**Reply by djspiewak** [reply]
> Idk. Seems to be just my opinion at this point 🙂 I am also disappointed that MUnit released for SN 0.5 without fixing any of the single-threaded assumptions baked in. It all feels very haphazard.

I agree this feels pretty haphazard. People seem to be in the mode of "let's unblock everyone first", and once that's done I suspect everyone will reconsider the multi-threaded implications. Unfortunately, I think relatively few people have thought through the API breaking effects to any depth, which in turn means that we're implicitly signing up for a breaking change mid-cycle on SN 0.5.

I'm not entirely sure how we want to navigate that. Very *likely* it's just going to be done as a minor version rev across the board and we'll handwave about the guarantees, effectively foisting the problem on downstream users.

**Reply by etspaceman** [reply]
Just casting a +1 to supporting 0.5 to unblock updates. I'm not sure if that has been decided, but I'd really like to not have to hold off on updates to items like munit until we figure out multi-threading.

**Reply by keynmol** [reply]
> and once that's done I suspect everyone will reconsider the multi-threaded implications

With the way our build tools work these days, it's just incredibly easier to work on your own library's broken tests once all the upstream is published. And most of the core dependencies have been pretty good with frequent publishing.

So at this point the goal is to solve the "my library is 10 levels deep with unreleased dependencies" problem. Solving fundamental blockers unblocks the maintainer parallelism in the ecosystem, and then the fun part starts with removing single-threaded assumptions from native-only code. Of which I can imagine there isn't actually a ton of in most libraries.

**Reply by ekrich** [reply]
After thinking about this for awhile I am also +1 here to unblock `0.5.x` because multi-thread is new anyway. This way there can be time for a future multi-thread release. Changes may be needed in `0.5.x` anyway to make things work properly.

**Comment by mpilquist** [comment]
We're *not* supporting native 0.5 on cats-effect 2, right?

**Reply by djspiewak** [reply]
Yeah definitely not. Unless you want to shave some yaks for absolutely no reason.

**Comment by hamnis** [comment]
Perhaps circe/circe should be added to the board?

https://github.com/circe/circe/pull/2254
